### DISCUSSION TITLE
Situations where maliciuos IPs get HTTP 401 reponses?

### DISCUSSION THREAD

**Discussion by intrapid** [OpenIDC/mod_auth_openidc#1119] [Q&A] [answered]
Hi!

I was wondering if there any situations where malicious IPs can get 401 responses (`"GET / HTTP/1.1" 401`) from Apache, when the entire site requires authentication with mod_auth_openidc?

Wouldn't all unauthenticated clients get `302` redirects to the openid provider? And that would show up in the apache log?

Site config:
```
<Location />
    SSLRequireSSL
    SSLOptions +StdEnvVars

    AuthType openid-connect
    Require claim XXX:YYYY

    Options Includes FollowSymLinks
    AllowOverride AuthConfig Limit
    Order allow,deny
    Allow from all
</Location>	
```

**Comment by zandbelt** [comment, marked as answer]
when a non-browser request is detected, a 401 would be returned, otherwise a 302 would happen indeed, regardless of the IP address used; if you want to block certain IP ranges you can do it through Apache configurations or through a dedicated component such as a web application firewall or ingress proxy

**Reply by intrapid** [reply]
Thanks! 

OK, so when someone is running a script that is probing the site, 401 responses are to be expected.

Is there a way to see the difference in the logs between these two scenarios?

* Unauthorized non-browser requests 
* Authorized (by the identity provider) browser requests where the `Require claim` is not fulfilled

**Reply by zandbelt** [reply]
you can see the difference in the error log when the LogLevel is set to debug, otherwise there's also an option to make it appear in the access logs since 2.4.13.2 because authorization errors are storered in environment variable `OIDC_AUTHZ_ERROR`
so its value can be used e.g. with HTTP 401 responses in the access log:
```
    LogFormat "%h %l %u %t %U %401{OIDC_AUTHZ_ERROR}e %>s %b" combined
```

**Reply by intrapid** [reply]
Excellent! I will incorporate that in the logs.

**Reply by monolied** [reply]
This is interesting. I imagine, there are a handful of criteria to determine "non-browser" - clients. One of them _might_ be the Accept-Header, I had to realize lately.
`curl <site> # 302`
`curl --header 'Accept: ' <site> # 401`
The 401 is, btw, also given to **authorized** clients in this situation. After reading this, I'm unsure if to file this as a bug.
The Accept header isn't mandatory and everything else regarding authorization/authentication works without, so it is at least, say, irritating.
Also the fact the log throws:
`[auth_openidc:warn] [pid 3978960:tid 139905506195200] [client x.x.x.x:59830] oidc_get_provider_from_session: empty or invalid session: no issuer found` seems to indicate more of a bug, and less of a feature.

**Reply by zandbelt** [reply]
see https://github.com/OpenIDC/mod_auth_openidc/blob/v2.4.14.3/auth_openidc.conf#L910-L935

**Reply by monolied** [reply]
Reasonable; but the error message then still seems to be quite off.

**Reply by zandbelt** [reply]
I cannot reproduce that log warning when using the request you provided. What version of mod_auth_openidc are you on?

**Reply by monolied** [reply]
Should be the latest Debian bullseye package; but I'll verify when I'm on site.

**Reply by monolied** [reply]
Version: 2.4.9.4-0+deb11u3. But, it seems now, while the error numbers for the request are given correctly, the error message was probably related to the logout following the failing request. And the ones related to the shortened curl request must have been rather unlucky coincidences with other requests on that live system so I thought they were a direct connection and reproducing the error. Insofar I have to apologize for the irritation about the apache log excerpt.

**Reply by Norr** [reply]
> when a non-browser request is detected, a 401 would be returned, otherwise a 302 would happen indeed, regardless of the IP address used; if you want to block certain IP ranges you can do it through Apache configurations or through a dedicated component such as a web application firewall or ingress proxy

So that is mean non-browser connection is not supported? For example if I want to connect to webdav via Microsoft built-in client it will not work, right?
